File: models/losses/velocity_gprop.jl
# Custom PySR loss for models/velocity_gprop.py — joint SR of f(x, γ) and
# γ = g(galaxy properties), v3.
#
# Population model: γ_i ~ LogNormal(g(obs_i), σ_int). Per galaxy the inner
# optimiser maximises likelihood × prior: γ confined to γ_pred·10^{±3σ_int}
# with the smooth penalty ((log10 γ − log10 γ_pred)/σ_int)² added to the NLL
# (negative log of the lognormal prior; replaces the v2 hard band).
#
# Inner optimisation (settings from analysis/diagnose_inner_opt.py, accuracy-
# gated: v2's 5×3 grid + margin-shrunk NM box misranked candidates, Spearman
# 0.83 → 1.00 after these fixes; this config keeps Spearman 1.00 at 2.6× the
# throughput of the maximal one):
#   - coarse grid: __N_START_D__ log-d × __N_START_GAMMA__ γ points at
#     __COARSE_IRLS__ IRLS iterations (cheap basin finding)
#   - Nelder-Mead restarts from the TOP-2 grid cells (multimodal surface),
#     full __N_IRLS__-IRLS objective; best coarse cell rescored at full IRLS
#   - NM bounds = FULL (d, γ) box — no interior margin (edge optima reachable)
#
# Constant-g guard: γ_pred is evaluated for every galaxy up front; if the std
# of the (clamped) predictions across galaxies falls below GVAR_MIN, a large
# smooth penalty (up to WEIGHT_GVAR) is added — g must genuinely vary with
# galaxy properties. (PySR-side constants are disabled — weight_optimize=0,
# complexity_of_constants=99 — rationals come from the one/two/three atoms;
# the guard is what stops g collapsing to a constant expression.)
#
# dataset.X rows: 1=x 2=γ 3=xpow 4=one 5=two 6=three 7=galaxy_code
#                 8=Vgas² 9=Vdisk² 10=Vbulge² 11..(10+N_PROPS)=g's property
#                 features (anchored logs; constant within a galaxy block)
#
# Hyperparameters appear as double-underscored UPPERCASE placeholder tokens,
# filled in by models/julia_losses.py.
using Optim, LinearAlgebra
function physicsloss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
    # Sterile-member guard: out-of-arity features read zero rows, can score
    # well (atan(x10)=0 is a free zero) yet all children are rejected by
    # check_constraints → sterile tournament winners freeze evolution. Same
    # for real constants while complexity_of_constants=99 bans them (any
    # constant-bearing child exceeds maxsize). Hard-fail both.
    bad_f = any(n -> n.degree == 0 && (n.constant || n.feature > 6), tree.trees.f.tree)
    bad_g = any(n -> n.degree == 0 && (n.constant || n.feature > __G_MAX_FEATURE__), tree.trees.g.tree)
    (bad_f || bad_g) && return L(1e12)

    n_pts      = length(dataset.y)
    n_gal      = Int(maximum(dataset.X[7, :]))
    total_loss = L(0)

    ν  = L(__NU_T__)

    upsilon_wt = L(__UPSILON_WEIGHT__)
    σ_ml       = L(0.1 * log(10))
    μ_disk     = log(L(0.5))
    μ_bulge    = log(L(0.7))

    gamma_lo   = T(__GAMMA_LO__)
    gamma_hi   = T(__GAMMA_HI__)
    σ_int      = T(__SIGMA_INT__)    # lognormal scatter of γ about g(obs), dex
    gr_wt      = L(__WEIGHT_GAMMA_RANGE__)

    n_props  = __N_PROPS__       # galaxy-property features (g's locals 1..n_props)
    n_feat   = __N_FEATURES__    # total dataset features = rows of eval matrices
    gvar_min = T(__GVAR_MIN__)   # min std(γ_pred) across galaxies
    gvar_wt  = L(__WEIGHT_GVAR__)

    three_sigma = T(3) * σ_int   # γ box half-width in dex

    # f eval matrices: f's LOCAL indices 1=x,2=gamma,3=xpow,4=one,5=two,6=three.
    # n_feat rows (= n dataset features): random init trees may reference any
    # feature before check_constraints prunes them; remaining rows stay zero.
    X_origin = zeros(T, n_feat, 1)
    X_origin[1, 1] = T(__ORIGIN_X__)
    X_origin[4, 1] = T(1)
    X_origin[5, 1] = T(2)
    X_origin[6, 1] = T(3)

    n_phys_vel = 6
    X_phys_vel = zeros(T, n_feat, n_phys_vel)
    X_phys_vel[1, :] = [T(0.01), T(0.1), T(0.5), T(1.0), T(3.0), T(10.0)]
    X_phys_vel[4, :] .= T(1)
    X_phys_vel[5, :] .= T(2)
    X_phys_vel[6, :] .= T(3)

    # g eval matrix: g's LOCAL indices 1..n_props = properties, then
    # n_props+1=one, n_props+2=two, n_props+3=three. n_feat rows as above.
    X_gp = zeros(T, n_feat, 1)
    X_gp[n_props + 1, 1] = T(1)
    X_gp[n_props + 2, 1] = T(2)
    X_gp[n_props + 3, 1] = T(3)

    # O(n_pts) galaxy index scan (data sorted by galaxy_code in Python).
    g_start = zeros(Int, n_gal)
    g_end   = zeros(Int, n_gal)
    for i in 1:n_pts
        g = Int(dataset.X[7, i])
        if g_start[g] == 0; g_start[g] = i; end
        g_end[g] = i
    end

    # ---- pre-pass: γ_pred = g(props) for every galaxy + constant-g guard ----
    γ_pred = fill(T(NaN), n_gal)
    for g in 1:n_gal
        g_start[g] == 0 && continue
        for p in 1:n_props
            X_gp[p, 1] = dataset.X[10 + p, g_start[g]]
        end
        g_vals, ok_g = eval_tree_array(tree.trees.g, X_gp, options)
        (!ok_g || !isfinite(g_vals[1])) && return L(1e12)
        γ_pred[g] = g_vals[1]
    end
    # std over CLAMPED predictions: g's that vary wildly but clamp to a
    # constant (e.g. 10·x_M → all γ_c = γ_hi) count as constant too.
    γc_all = [clamp(γ_pred[g], gamma_lo + T(1e-3), gamma_hi - T(1e-3))
              for g in 1:n_gal if g_start[g] != 0]
    n_ok = length(γc_all)
    n_ok == 0 && return L(1e12)
    μ_γ = sum(γc_all) / n_ok
    σ_γ = sqrt(sum(abs2, γc_all .- μ_γ) / n_ok)
    gvar_pen = σ_γ < gvar_min ? gvar_wt * L(1 - σ_γ / gvar_min)^2 : L(0)

    for g in 1:n_gal
        g_start[g] == 0 && continue
        n_g = g_end[g] - g_start[g] + 1
        n_g < 3 && continue

        # ---- γ prior from galaxy properties (pre-pass above) ----
        γ_raw = γ_pred[g]
        γ_c   = clamp(γ_raw, gamma_lo + T(1e-3), gamma_hi - T(1e-3))
        range_pen = min(gr_wt * L(γ_raw - γ_c)^2, L(1e6))
        lg_γc = log10(γ_c)

        # γ box: γ_c·10^{±3σ_int} ∩ (gamma_lo, gamma_hi); prior penalty inside
        # loss_g keeps the box interior soft (max penalty at edge = 9).
        γ_box_lo = max(gamma_lo + T(1e-3), γ_c * exp10(-three_sigma))
        γ_box_hi = min(gamma_hi - T(1e-3), γ_c * exp10(three_sigma))

        rng_g    = g_start[g]:g_end[g]
        r_g      = view(dataset.X, 1, rng_g)
        Vgas2_g  = view(dataset.X, 8, rng_g)
        Vdisk2_g = view(dataset.X, 9, rng_g)
        Vbul2_g  = view(dataset.X, 10, rng_g)
        y_g      = view(dataset.y, rng_g)
        w_g      = view(dataset.weights, rng_g)
        resid_g  = y_g .- Vgas2_g

        W_sqrt  = sqrt.(w_g)
        resid_w = W_sqrt .* resid_g

        has_bulge_g = maximum(abs, Vbul2_g) > T(1e-6)

        # Per-galaxy adaptive d range.
        r_max_g    = maximum(r_g)
        d_min_g    = r_max_g / T(20)
        d_max_g    = max(r_max_g * T(10), T(50))
        log_d_lo_g = log(d_min_g)
        log_d_hi_g = log(d_max_g)

        best_g     = L(Inf)
        best_log_d = (log_d_lo_g + log_d_hi_g) / T(2)
        best_gamma = γ_c

        # ---- Pre-allocate work arrays once per galaxy (outside loss_g). ----
        x_vals      = Vector{T}(undef, n_g)
        xpow_buf    = Vector{T}(undef, n_g)
        X_loc       = zeros(T, n_feat, n_g)
        A           = Matrix{T}(undef, n_g, 3)
        A_w         = Matrix{T}(undef, n_g, 3)
        A_irls      = Matrix{T}(undef, n_g, 3)
        r_work      = Vector{T}(undef, n_g)
        w_irls_buf  = Vector{T}(undef, n_g)
        W_irls_sqrt = Vector{T}(undef, n_g)
        resid_irls  = Vector{T}(undef, n_g)
        coefs_buf   = zeros(T, 3)
        best_coefs  = zeros(T, 3)

        # Fixed columns: f's local indices 4=one, 5=two, 6=three.
        X_loc[4, :] .= T(1)  # one
        X_loc[5, :] .= T(2)  # two
        X_loc[6, :] .= T(3)  # three
        A[:, 1]      = Vdisk2_g
        A[:, 2]      = Vbul2_g
        A_w[:, 1]    = W_sqrt .* Vdisk2_g
        A_w[:, 2]    = W_sqrt .* Vbul2_g

        # ---- Loss closure: NLL + γ prior; mutates preallocated buffers. ----
        function loss_g(params::AbstractVector{T}, n_irls_local::Int=__N_IRLS__)
            log_d_p, γ_p = params[1], params[2]
            d_val = exp(log_d_p)
            @. x_vals   = r_g / d_val
            @. xpow_buf = clamp(x_vals, T(1e-3), T(1e3)) ^ (-γ_p)
            X_loc[1, :] .= x_vals    # x     (local index 1)
            X_loc[2, :] .= γ_p      # gamma (local index 2)
            X_loc[3, :] .= xpow_buf # xpow  (local index 3)

            f_vals, flag = eval_tree_array(tree.trees.f, X_loc, options)
            !flag && return L(1e12)
            any(!isfinite, f_vals) && return L(1e12)
            maximum(abs, f_vals) < T(1e-6) && return L(1e12)

            A[:, 3]   .= f_vals
            A_w[:, 3] .= W_sqrt .* f_vals
            maximum(abs, A) < T(1e-10) && return L(1e12)

            coefs = try; A_w \ resid_w; catch; return L(1e12); end
            coefs = max.(coefs, zero(T))

            for irls_iter in 1:n_irls_local
                r_work .= A * coefs
                r_work .-= resid_g
                @. w_irls_buf = w_g * (ν + one(ν)) / (ν + w_g * r_work ^ 2)
                W_irls_sqrt .= sqrt.(w_irls_buf)
                A_irls[:, 1] .= W_irls_sqrt .* Vdisk2_g
                A_irls[:, 2] .= W_irls_sqrt .* Vbul2_g
                A_irls[:, 3] .= W_irls_sqrt .* f_vals
                resid_irls   .= W_irls_sqrt .* resid_g
                new_coefs = try; A_irls \ resid_irls; catch; break; end
                any(!isfinite, new_coefs) && break
                coefs = max.(new_coefs, zero(T))
            end

            r_work .= A * coefs
            r_work .-= resid_g
            st = sum(log(T(1) + w_g[i] * r_work[i]^2 / ν) for i in eachindex(r_work)) / n_g
            (isinf(st) || isnan(st)) && return L(1e12)

            hinge = T(1000) * (max(-coefs[1], zero(T))^2 +
                               max(-coefs[2], zero(T))^2 +
                               max(-coefs[3], zero(T))^2)
            prior = zero(T)
            if coefs[1] > T(1e-30)
                prior += (log(coefs[1]) - μ_disk)^2 / (2 * σ_ml^2)
            end
            if has_bulge_g && coefs[2] > T(1e-30)
                prior += (log(coefs[2]) - μ_bulge)^2 / (2 * σ_ml^2)
            end
            # lognormal γ prior: negative log of LogNormal(γ_c, σ_int)
            prior_γ = ((log10(γ_p) - lg_γc) / σ_int)^2
            coefs_buf .= coefs
            return st + hinge + upsilon_wt * prior + prior_γ
        end

        # ---- Coarse grid at __COARSE_IRLS__ IRLS (cheap basin find), top-2 ----
        v_s1 = L(Inf); d_s1 = best_log_d; g_s1 = γ_c
        v_s2 = L(Inf); d_s2 = best_log_d; g_s2 = γ_c
        for gamma_j in range(γ_box_lo, γ_box_hi; length=__N_START_GAMMA__)
            for log_d_j in range(log_d_lo_g, log_d_hi_g; length=__N_START_D__)
                c = loss_g(T[log_d_j, gamma_j], __COARSE_IRLS__)
                if c < v_s1
                    v_s2 = v_s1; d_s2 = d_s1; g_s2 = g_s1
                    v_s1 = c;    d_s1 = log_d_j; g_s1 = gamma_j
                elseif c < v_s2
                    v_s2 = c;    d_s2 = log_d_j; g_s2 = gamma_j
                end
            end
        end
        # rescore best coarse cell at the full-IRLS objective
        if isfinite(v_s1)
            best_g = loss_g(T[d_s1, g_s1])
            best_coefs .= coefs_buf
            best_log_d = d_s1
            best_gamma = g_s1
        end

        # ---- Nelder-Mead (Fminbox) restarts from the top-2 coarse cells ----
        # FULL box — no interior margin (v2's d_step/2 shrink excluded edge optima).
        lb_box = T[log_d_lo_g, γ_box_lo]
        ub_box = T[log_d_hi_g, γ_box_hi]
        if isfinite(best_g)
            ϵd = T(1e-6) * (log_d_hi_g - log_d_lo_g)
            ϵg = T(1e-6) * max(γ_box_hi - γ_box_lo, T(1e-6))
            for (d0, g0, v0) in ((d_s1, g_s1, v_s1), (d_s2, g_s2, v_s2))
                isfinite(v0) || continue
                p0_int = T[clamp(d0, lb_box[1] + ϵd, ub_box[1] - ϵd),
                           clamp(g0, lb_box[2] + ϵg, ub_box[2] - ϵg)]
                res = try
                    Optim.optimize(
                        loss_g, lb_box, ub_box, p0_int,
                        Optim.Fminbox(Optim.NelderMead()),
                        Optim.Options(iterations=__OPTIMIZER_NITER__, f_reltol=L(1e-7)))
                catch
                    continue
                end
                v_r = Optim.minimum(res)
                if v_r < best_g
                    best_g     = v_r
                    best_log_d = T(Optim.minimizer(res)[1])
                    best_gamma = T(Optim.minimizer(res)[2])
                    loss_g(T.(Optim.minimizer(res)))
                    best_coefs .= coefs_buf
                end
            end
        end

        # ---- Use c saved at best loss ----
        best_c = best_coefs[3]

        # ---- origin penalty: enforce f(0, γ) ≈ 0 at optimal γ ----
        if L(__ORIGIN_WEIGHT__) > L(0)
            X_origin[2, 1] = best_gamma
            X_origin[3, 1] = T(__ORIGIN_X__) ^ (-best_gamma)
            f_orig, ok_orig = eval_tree_array(tree.trees.f, X_origin, options)
            if ok_orig && isfinite(f_orig[1])
                best_g += L(__ORIGIN_WEIGHT__) * min((best_c * f_orig[1])^2, L(1000))
            else
                best_g += L(__ORIGIN_WEIGHT__) * L(1000)
            end
        end

        # ---- positivity and mass-monotonicity penalties ----
        if L(__WEIGHT_NONNEG__) > L(0) || L(__WEIGHT_MONO__) > L(0)
            X_phys_vel[2, :] .= best_gamma
            X_phys_vel[3, :] .= X_phys_vel[1, :] .^ (-best_gamma)
            f_phys, ok_fphys = eval_tree_array(tree.trees.f, X_phys_vel, options)
            if ok_fphys && all(isfinite, f_phys)
                if L(__WEIGHT_NONNEG__) > L(0)
                    nonneg_pen = sum(max(-(best_c * v), T(0))^2 for v in f_phys)
                    best_g += L(__WEIGHT_NONNEG__) * min(nonneg_pen, L(1e6))
                end
                if L(__WEIGHT_MONO__) > L(0)
                    xf = X_phys_vel[1, :] .* (best_c .* f_phys)
                    mono_pen = sum(max(xf[k] - xf[k+1], T(0))^2 for k in 1:length(xf)-1)
                    best_g += L(__WEIGHT_MONO__) * min(mono_pen, L(1e6))
                end
            end
        end

        isinf(best_g) && (best_g = L(1e12))
        best_g += range_pen
        total_loss += best_g
    end

    return total_loss / n_gal + gvar_pen
end

File: models/velocity_gprop.py
import json
import os

from pysr import PySRRegressor, TemplateExpressionSpec
import pandas as pd
import numpy as np

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from models.julia_losses import load_loss
from models.gprop_features import galaxy_features, PROP_NAMES, N_PROPS


def fit_vr_gprop(df: pd.DataFrame,
                 output_directory: str = "outputs",
                 params_csv: str = "data/SPARC/galaxy_parameters.csv",
                 error_weighting: bool = True,
                 iterations: int = 99999,
                 n_galaxies: int | None = 5,
                 n_start_d: int = 12,
                 n_start_gamma: int = 5,
                 coarse_irls: int = 1,
                 gamma_range: tuple = (0.0, 1.0),
                 sigma_int: float = 0.05,
                 weight_gamma_range: float = 1.0,
                 gvar_min: float = 0.05,
                 weight_gvar: float = 100.0,
                 populations: int = 20,
                 population_size: int = 40,
                 ncycles_per_iteration: int = 200,
                 weight_optimize: float = 0.0,
                 optimizer_iterations: int = 8,
                 upsilon_weight: float = 1.0,
                 origin_weight: float = 1.0,
                 origin_x: float = 1e-3,
                 weight_nonneg: float = 0.1,
                 weight_mono: float = 0.1,
                 binary_operators: list = ["/", "-", "+", "*"],
                 nu_t: float = 3.0,
                 n_irls: int = 5,
                 min_points: int = 5,
                 unary_operators: list = None,
                 maxsize: int = 30,
                 optimizer_niter: int = 100,
                 timeout_in_seconds: float | None = None,
                 guesses: list | None = None,
                 procs: int = 0,
                 run_id: str | None = None,
                 warm_start: bool = False,
                 checkpoint_every: int | None = None):
    """
    Joint SR: learn f(r/d, γ) AND γ = g(galaxy properties) simultaneously.

    Model (per data point i in galaxy g), v3:
        V²_obs = sign(Vgas)·V²_gas + a[g]·V²_disk + b[g]·V²_bul + c[g]·f(r/d[g], γ[g])
        γ[g]   ~ LogNormal( g(x_M, x_f, x_S, x_R)[g], sigma_int )

    The lognormal population prior is implemented in the loss: per galaxy, γ is
    optimised inside γ_pred·10^{±3·sigma_int} with the smooth penalty
    ((log10 γ − log10 γ_pred)/sigma_int)² added to the NLL (replaces the v2
    hard band). Features come from models/gprop_features.py (photometric/gas
    only, fixed anchors — see that module).

    Per-galaxy parameters:
      a, b — upsilon_disk, upsilon_bulge (linear, WLS → Student-t IRLS solve)
      c    — DM amplitude (linear)
      d    — DM scale radius (nonlinear; per-galaxy adaptive range)
      γ    — within the lognormal prior around g(props)
             (NOT PySR parameters — avoids SVector{N} compile bottleneck)

    Inner-opt settings (n_start_d=12, n_start_gamma=5 within the prior box at
    coarse_irls=1, top-2 Nelder-Mead restarts on the FULL box, n_irls=5) are
    accuracy-gated by analysis/diagnose_inner_opt.py — v2's 5×3 grid +
    margin-shrunk NM box misranked candidates (Spearman 0.83 → 1.00 with these
    settings, at 2.6× the throughput of the maximal-accuracy config).

    PySR-side constants are DISABLED (weight_optimize=0, complexity_of_constants
    =99, should_optimize_constants=False — the velocity2param recipe): BFGS
    constant tuning nests full inner optimisations inside finite-difference
    gradients (dozens of full-dataset evals per call) and starved the search.
    Rational coefficients are built from the one/two/three atoms instead
    (e.g. 1/2 − x_M/5 = #5/#6 − #1/(#6+#7)); seeds carry such forms.

    Guards (in models/losses/velocity_gprop.jl):
      - sterile-member guard (arity only): hard-fail out-of-arity features
      - constant-g guard: std of clamped γ_pred across galaxies must exceed
        gvar_min, else a smooth penalty up to weight_gvar — g must genuinely
        vary with galaxy properties

    Sub-expression feature indices are LOCAL to each combine argument list
    (check_constraints rejects trees with g features >n_props+3 / f features >6):
      f(x, ·, xpow, one, two, three): local 1=x, 2=gamma, 3=xpow, 4=one, 5=two, 6=three
      g(<props>, one, two, three):    local 1..n_props = props, then one, two, three
    BUT randomly initialised trees may reference any dataset feature before
    constraint checks prune them, so all eval matrices must have n_features rows
    (extra rows zero → garbage loss → pruned).
    dataset.X layout (10 + n_props rows):
      rows 1-6:   x, γ, xpow, one, two, three (f's args; 1-3 overwritten per-galaxy)
      rows 7-10:  galaxy_code, Vgas2, Vdisk2, Vbulge2 (aux)
      rows 11+:   anchored galaxy properties (constant within galaxy block;
                  fixed a-priori anchors, no sample statistics — see PROPS below)
    """
    if n_galaxies is not None:
        available = df["galaxy"].unique()
        rng = np.random.default_rng(seed=42)
        selected = rng.choice(available, size=min(n_galaxies, len(available)), replace=False)
        df = df[df["galaxy"].isin(selected)].copy()

    counts = df.groupby("galaxy")["galaxy"].transform("count")
    df = df[counts >= min_points].copy()

    # ---- Galaxy properties (g's features) — see models/gprop_features.py ----
    from models.gprop_features import PROPS
    n_props = N_PROPS
    prop_names = list(PROP_NAMES)
    props = galaxy_features(pd.read_csv(params_csv))

    n_before = df["galaxy"].nunique()
    df = df.drop(columns=[c for c in ("log_Mstar", "log_fgas", *prop_names)
                          if c in df.columns])
    df = df.merge(props, on="galaxy", how="inner")
    n_after = df["galaxy"].nunique()
    if n_after < n_before:
        print(f"Dropped {n_before - n_after} galaxies missing properties in {params_csv}")

    scaler = dict(PROPS)

    os.makedirs(output_directory, exist_ok=True)
    with open(os.path.join(output_directory, "feature_scaler.json"), "w") as fh:
        json.dump(scaler, fh, indent=2)

    galaxy_codes, galaxies = pd.factorize(df["galaxy"])
    df["galaxy_code"] = galaxy_codes + 1
    df = df.sort_values("galaxy_code").reset_index(drop=True)

    y = df["Vobs_km/s"] ** 2
    X = pd.DataFrame({
        "r":           df["Rad_kpc"],           # global 1 = x
        "gamma_dummy": np.zeros(len(df)),       # global 2 = gamma  (overwritten per-galaxy in Julia)
        "xpow_dummy":  np.zeros(len(df)),       # global 3 = xpow   (overwritten per-galaxy in Julia)
        "one":         np.ones(len(df)),        # global 4 = one
        "two":         np.full(len(df), 2.0),   # global 5 = two
        "three":       np.full(len(df), 3.0),   # global 6 = three
        "galaxy":      df["galaxy_code"],       # global 7 — galaxy index scan
        "Vgas2":       np.sign(df["Vgas_km/s"]) * df["Vgas_km/s"] ** 2,  # global 8
        "Vdisk2":      df["Vdisk_km/s"] ** 2,   # global 9
        "Vbulge2":     df["Vbul_km/s"] ** 2,    # global 10
        # globals 11..(10+n_props) — g's locals 1..n_props
        **{name: df[name] for name in prop_names},
    })

    gamma_lo, gamma_hi = gamma_range

    # Custom loss (models/losses/velocity_gprop.jl): γ ~ LogNormal(g(props),
    # sigma_int); (a, b, c) analytic, (log d, γ) by dense coarse grid + top-3
    # Fminbox Nelder-Mead restarts; sterile-arity and constant-g guards.
    inner_opt_loss = load_loss(
        "velocity_gprop",
        nu_t=nu_t,
        upsilon_weight=upsilon_weight,
        gamma_lo=gamma_lo,
        gamma_hi=gamma_hi,
        sigma_int=sigma_int,
        weight_gamma_range=weight_gamma_range,
        n_props=n_props,
        n_features=10 + n_props,
        g_max_feature=n_props + 3,
        gvar_min=gvar_min,
        weight_gvar=weight_gvar,
        origin_x=origin_x,
        n_irls=n_irls,
        coarse_irls=coarse_irls,
        n_start_d=n_start_d,
        n_start_gamma=n_start_gamma,
        optimizer_niter=optimizer_niter,
        origin_weight=origin_weight,
        weight_nonneg=weight_nonneg,
        weight_mono=weight_mono,
    )

    template = TemplateExpressionSpec(
        expressions=["f", "g"],
        variable_names=["x", "gamma", "xpow", "one", "two", "three",
                        "galaxy_code", "Vgas2", "Vdisk2", "Vbulge2",
                        *prop_names],
        combine=f"f(x, g({', '.join(prop_names)}, one, two, three), "
                "xpow, one, two, three)",
    )

    # Seed populations with feature-valid members. Random init trees sample all
    # dataset features, but valid members need f-features ≤6 AND g-features
    # ≤n_props+3 (joint check_constraints), so unseeded populations start
    # ~all-invalid and take a long time to bootstrap.
    # f seeds: physically motivated profiles (see profiles.py) — pISO, NFW,
    # SR-A (C11), SR-C9, SR-G23. g seeds: linear laws along the known post-hoc
    # relation (γ decreasing with M*, slope ≈ −0.2) + variants. PySR constants
    # are disabled, so coefficients are rationals built from the atoms.
    # Placeholders are LOCAL: f: #1=x #2=γ #3=xpow=x^(−γ) #4=1 #5=2 #6=3;
    #   g: #1=x_M #2=x_f #3=x_S #4=x_R #5=1 #6=2 #7=3.
    if guesses is None:
        g_lin_M = "#5/#6 - #1/(#6 + #7)"        # 1/2 − x_M/5 ≈ known law
        g_lin_F = "#5/#7 + #2/(#6 + #7)"        # 1/3 + x_f/5
        g_lin_S = "#5/#6 - #3/(#6 * #7)"        # 1/2 − x_S/6
        g_lin_MS = "#5/#6 - (#1 + #3)/(#6 * #7)"  # 1/2 − (x_M+x_S)/6
        guesses = [
            {"f": "#4 - atan(#1)/#1",              "g": g_lin_M},   # pISO + M* law
            {"f": "#4 - atan(#1)/#1",              "g": g_lin_S},   # pISO + Σ* law
            {"f": "(log1p(#1) - #1/(#4+#1))/#1",   "g": g_lin_M},   # NFW
            {"f": "#3*atan(#6*log1p((#1/#3)/#3))", "g": g_lin_M},   # SR-A (C11)
            {"f": "#3*atan(#6*log1p((#1/#3)/#3))", "g": g_lin_MS},  # SR-A + M,Σ
            {"f": "#1/(#1 + #3 - #2*#6)",          "g": g_lin_F},   # SR-C9 + fgas
            {"f": "atan(#3*(#1 - log1p(#1))/(#2*(#1 + log1p(#6))))",
                                                    "g": g_lin_M},  # SR-G19 family
        ]

    active_unary = unary_operators if unary_operators is not None else ["atan", "log1p"]
    nested = {op: {op: 0} for op in active_unary}
    if "exp" in active_unary:
        nested["exp"]["log1p"] = 0  # exp(log1p(x)) ≈ x, trivial
        if "log" in active_unary:
            nested["exp"]["log"] = 0   # exp(log(x)) = x, trivial
            nested["log"]["exp"] = 0   # log(exp(x)) = x, trivial
        if "log1p" in active_unary:
            nested["log1p"]["exp"] = 0 # log1p(exp(x)) ≈ x for large x
    # Resume from a prior checkpoint (SL3-chained run) if one exists; otherwise
    # build fresh. run_id must be fixed (not the PySR-default timestamp) for
    # the checkpoint path to be findable across separate sbatch submissions.
    checkpoint = Path(output_directory) / str(run_id) / "checkpoint.pkl" if run_id else None
    if warm_start and checkpoint is not None and checkpoint.exists():
        # Not PySRRegressor.from_file(): that eagerly calls model.refresh(),
        # which reconstructs the hall of fame from Julia state that isn't
        # valid for a run interrupted mid-search (crashes with "cannot unpack
        # non-iterable NoneType" in get_hof()). The actual resume mechanism
        # lives in .fit() itself (keyed on warm_start + julia_state_stream_,
        # sr.py ~line 2458) and doesn't need refresh() first, so just unpickle
        # and let fit() below do the resume.
        import pickle
        print(f"[resume] loading checkpoint from {checkpoint} (skipping PySR's own refresh())")
        with open(checkpoint, "rb") as f:
            model = pickle.load(f)
        model.set_params(warm_start=True)
    else:
        model = PySRRegressor(
            expression_spec=template,
            output_directory=output_directory,
            run_id=run_id,
            warm_start=warm_start,
            niterations=iterations,
            binary_operators=binary_operators,
            unary_operators=active_unary,
            nested_constraints=nested,
            maxsize=maxsize,
            populations=populations,
            population_size=population_size,
            ncycles_per_iteration=ncycles_per_iteration,
            weight_optimize=weight_optimize,
            optimizer_iterations=optimizer_iterations,
            should_optimize_constants=False,
            complexity_of_constants=99,
            batching=False,
            turbo=True,
            timeout_in_seconds=timeout_in_seconds,
            procs=procs,
            guesses=guesses,
            loss_function_expression=inner_opt_loss,
        )

    if error_weighting:
        errV_safe = np.maximum(df["errV_km/s"], 0.5)
        weights = 1.0 / (2 * df["Vobs_km/s"] * errV_safe) ** 2
    else:
        weights = np.ones(len(df))

    if checkpoint_every is not None:
        # PySR only checkpoints before _run() starts and after it returns
        # normally (sr.py fit()) — a niterations=99999 run killed by SLURM's
        # wall-clock never reaches the "after" checkpoint, so nothing from
        # this job's search survives to disk. Looping fit() in small chunks
        # means every chunk completes normally, so a real, resumable
        # checkpoint is written every `checkpoint_every` iterations instead
        # of only (never) at the end.
        model.set_params(niterations=checkpoint_every, warm_start=True)
        while True:
            model.fit(X, y, weights=weights)
    else:
        model.fit(X, y, weights=weights)


if __name__ == "__main__":
    from datawrangling import produce_SPARC_df
    df = produce_SPARC_df("data/SPARC", selected=True)
    fit_vr_gprop(
        df,
        output_directory="outputs/toydatasets/velocity_gprop_smoke",
        iterations=99999,
        n_galaxies=5,
        populations=10,
        population_size=20,
        ncycles_per_iteration=100,
        unary_operators=["atan", "log1p"],
        maxsize=28,
    )
